src/openms/include/OpenMS/ANALYSIS/ID/IDBoostGraph.h

```diff
@@ -0,0 +1,278 @@
+      {
+        std::cout << "set score " << posterior << " for " << prot->getAccession() << std::endl;
+        prot->setScore(posterior);
+      }
+
+      void operator()(ProteinGroup* /*protgrp*/, double /*posterior*/) const
+      {
+        // do nothing
+        //protgrp->probability = posterior;
+      }
+
+      void operator()(const PeptideCluster* /*pc*/, double /*posterior*/) const
+      {
+        // do nothing
+      }
+
+    };
+
+    /// Visits nodes in the boost graph (ptrs to an ID Object) and depending on their type creates a random
+    /// variable or "Dependency" to add into the InferenceGraph
+    /*class DependencyVisitor:
+        public boost::static_visitor<Dependency>
+    {
+    public:
+      const MessagePasserFactory& mpf;
+
+      DependencyVisitor(const MessagePasserFactory& mpf):
+      mpf(mpf)
+      {}
+
+      Dependency operator()(const PeptideHit* pep, const std::vector<vertex_t>& neighbors, const FilteredGraph& fg) const
+      {
+        std::vector<IDBoostGraph::vertex_t> incoming{};
+        for (const auto& nb : neighbors)
+        {
+          if (fg[nb].which() <= 2)
+          {
+            incoming.push_back(nb);
+          }
+        }
+        if (incoming.size() != 1)
+        {
+          std::cerr << "Incoming nodes for pep are more than 1. Sth went wrong." << std::endl;
+        }
+        return mpf.createPeptideEvidenceFactor()
+      }
+
+      Dependency operator()(const ProteinHit* prot, const std::vector<vertex_t>& neighbors, const FilteredGraph& fg) const
+      {
+        return prot->getAccession();
+      }
+
+      Dependency operator()(const PeptideCluster* pc, const std::vector<vertex_t>& neighbors, const FilteredGraph& fg) const
+      {
+        return pep->getSequence().toUnmodifiedString();
+      }
+
+      Dependency operator()(const ProteinGroup* pg, const std::vector<vertex_t>& neighbors, const FilteredGraph& fg) const
+      {
+        return prot->getAccession();
+      }
+
+    };*/
+
+    /// Compute connected component on the static graph. Needs to be recomputed if graph is changed.
+    void computeConnectedComponents();
+
+    /// Initialize and store the graph
+    /// IMPORTANT: Once the graph is built, editing members like (protein/peptide)_hits_ will invalidate it!
+    /// @param protein ProteinIdentification object storing IDs and groups
+    /// @param idedSpectra vector of ProteinIdentifications with links to the proteins and PSMs in its PeptideHits
+    /// @param use_all_psms If all or just the FIRST psm should be used
+    void buildGraph(bool use_all_psms);
+    //void buildGraph(const ProteinIdentification& protein, const std::vector<PeptideIdentification>& peptides);
+
+  private:
+    Graph g;
+
+    /// static objects created from a hard typedef to differentiate between different types of nodes.
+    /// they will be reused throughout the graph when necessary
+    static PeptideCluster staticPC;
```

Comment:
are these needed?

Reply:
Yes they are placeholders for intermediate nodes that connect proteins and peptides that share computations. Not used for basic inference/grouping.

Reply:
Can we these be non static?

Reply:
Also, I will probably change it in the future because I want to store info in the groups at least.

Reply:
if these are read-only then you could keep them but it must be static const.
If they are also written to, then they should not be static members as this will kill any concurrency (e.g., if multiple instances exist).